File: superset-backend/tests/unit_tests/dao/queries_test.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
from datetime import datetime, timedelta
from typing import Any

import pytest
from pytest_mock import MockerFixture
from sqlalchemy.orm.session import Session

from superset.exceptions import QueryNotFoundException, SupersetCancelQueryException


def test_query_dao_save_metadata(session: Session) -> None:
    from superset import db
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
    )

    db.session.add(database)
    db.session.add(query_obj)

    from superset.daos.query import QueryDAO

    query = db.session.query(Query).one()
    QueryDAO.save_metadata(query=query, payload={"columns": []})
    assert query.extra.get("columns", None) == []


def test_query_dao_get_queries_changed_after(session: Session) -> None:
    from superset import db
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    now = datetime.utcnow()

    old_query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        changed_on=now - timedelta(days=3),
    )

    updated_query_obj = Query(
        client_id="updated_foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from foo",
        select_sql="select * from foo",
        executed_sql="select * from foo",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        changed_on=now - timedelta(days=1),
    )

    db.session.add(database)
    db.session.add(old_query_obj)
    db.session.add(updated_query_obj)

    from superset.daos.query import QueryDAO

    timestamp = datetime.timestamp(now - timedelta(days=2)) * 1000
    result = QueryDAO.get_queries_changed_after(timestamp)
    assert len(result) == 1
    assert result[0].client_id == "updated_foo"


def test_query_dao_stop_query_not_found(
    mocker: MockerFixture, app: Any, session: Session
) -> None:
    from superset import db
    from superset.common.db_query_status import QueryStatus
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        status=QueryStatus.RUNNING,
    )

    db.session.add(database)
    db.session.add(query_obj)

    mocker.patch("superset.sql_lab.cancel_query", return_value=False)

    from superset.daos.query import QueryDAO

    with pytest.raises(QueryNotFoundException):
        QueryDAO.stop_query("foo2")

    query = db.session.query(Query).one()
    assert query.status == QueryStatus.RUNNING


def test_query_dao_stop_query_not_running(
    mocker: MockerFixture, app: Any, session: Session
) -> None:
    from superset import db
    from superset.common.db_query_status import QueryStatus
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        status=QueryStatus.FAILED,
    )

    db.session.add(database)
    db.session.add(query_obj)

    from superset.daos.query import QueryDAO

    QueryDAO.stop_query(query_obj.client_id)
    query = db.session.query(Query).one()
    assert query.status == QueryStatus.FAILED


def test_query_dao_stop_query_failed(
    mocker: MockerFixture, app: Any, session: Session
) -> None:
    from superset import db
    from superset.common.db_query_status import QueryStatus
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        status=QueryStatus.RUNNING,
    )

    db.session.add(database)
    db.session.add(query_obj)

    mocker.patch("superset.sql_lab.cancel_query", return_value=False)

    from superset.daos.query import QueryDAO

    with pytest.raises(SupersetCancelQueryException):
        QueryDAO.stop_query(query_obj.client_id)

    query = db.session.query(Query).one()
    assert query.status == QueryStatus.RUNNING


def test_query_dao_stop_query(
    mocker: MockerFixture, app: Any, session: Session
) -> None:
    from superset import db
    from superset.common.db_query_status import QueryStatus
    from superset.models.core import Database
    from superset.models.sql_lab import Query

    engine = db.session.get_bind()
    Query.metadata.create_all(engine)  # pylint: disable=no-member

    database = Database(database_name="my_database", sqlalchemy_uri="sqlite://")

    query_obj = Query(
        client_id="foo",
        database=database,
        tab_name="test_tab",
        sql_editor_id="test_editor_id",
        sql="select * from bar",
        select_sql="select * from bar",
        executed_sql="select * from bar",
        limit=100,
        select_as_cta=False,
        rows=100,
        error_message="none",
        results_key="abc",
        status=QueryStatus.RUNNING,
    )

    db.session.add(database)
    db.session.add(query_obj)

    mocker.patch("superset.sql_lab.cancel_query", return_value=True)

    from superset.daos.query import QueryDAO

    QueryDAO.stop_query(query_obj.client_id)
    query = db.session.query(Query).one()
    assert query.status == QueryStatus.STOPPED

File: superset-frontend/src/hooks/apiResources/queries.test.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import rison from 'rison';
import fetchMock from 'fetch-mock';
import { act, renderHook } from '@testing-library/react-hooks';
import {
  createWrapper,
  defaultStore as store,
} from 'spec/helpers/testing-library';
import { QueryState } from '@superset-ui/core';
import { api } from 'src/hooks/apiResources/queryApi';
import { mapQueryResponse, useEditorQueriesQuery } from './queries';

const fakeApiResult = {
  count: 4,
  ids: [692],
  result: [
    {
      changed_on: '2024-03-12T20:01:02.497775',
      client_id: 'b0ZDzRYzn',
      database: {
        database_name: 'examples',
        id: 1,
      },
      end_time: '1710273662496.047852',
      error_message: null,
      executed_sql: 'SELECT * from "FCC 2018 Survey"\nLIMIT 1001',
      id: 692,
      limit: 1000,
      limiting_factor: 'DROPDOWN',
      progress: 100,
      results_key: null,
      rows: 1000,
      schema: 'main',
      select_as_cta: false,
      sql: 'SELECT * from "FCC 2018 Survey" ',
      sql_editor_id: '22',
      start_time: '1710273662445.992920',
      status: QueryState.Success,
      tab_name: 'Untitled Query 16',
      tmp_table_name: null,
      tracking_url: null,
      user: {
        first_name: 'admin',
        id: 1,
        last_name: 'user',
      },
    },
  ],
};

afterEach(() => {
  fetchMock.reset();
  act(() => {
    store.dispatch(api.util.resetApiState());
  });
});

test('returns api response mapping camelCase keys', async () => {
  const editorId = '23';
  const editorQueryApiRoute = `glob:*/api/v1/query/?q=*`;
  fetchMock.get(editorQueryApiRoute, fakeApiResult);
  const { result, waitFor } = renderHook(
    () => useEditorQueriesQuery({ editorId }),
    {
      wrapper: createWrapper({
        useRedux: true,
        store,
      }),
    },
  );
  await waitFor(() =>
    expect(fetchMock.calls(editorQueryApiRoute).length).toBe(1),
  );
  const expectedResult = {
    ...fakeApiResult,
    result: fakeApiResult.result.map(mapQueryResponse),
  };
  expect(
    rison.decode(fetchMock.calls(editorQueryApiRoute)[0][0].split('?q=')[1]),
  ).toEqual(
    expect.objectContaining({
      order_column: 'start_time',
      order_direction: 'desc',
      page: 0,
      page_size: 25,
      filters: [
        {
          col: 'sql_editor_id',
          opr: 'eq',
          value: expect.stringContaining(editorId),
        },
      ],
    }),
  );
  expect(result.current.data).toEqual(expectedResult);
});

test('merges paginated results', async () => {
  const editorId = '23';
  const editorQueryApiRoute = `glob:*/api/v1/query/?q=*`;
  fetchMock.get(editorQueryApiRoute, fakeApiResult);
  const { waitFor } = renderHook(() => useEditorQueriesQuery({ editorId }), {
    wrapper: createWrapper({
      useRedux: true,
      store,
    }),
  });
  await waitFor(() =>
    expect(fetchMock.calls(editorQueryApiRoute).length).toBe(1),
  );
  const { result: paginatedResult } = renderHook(
    () => useEditorQueriesQuery({ editorId, pageIndex: 1 }),
    {
      wrapper: createWrapper({
        useRedux: true,
        store,
      }),
    },
  );
  await waitFor(() =>
    expect(fetchMock.calls(editorQueryApiRoute).length).toBe(2),
  );
  expect(
    rison.decode(fetchMock.calls(editorQueryApiRoute)[1][0].split('?q=')[1]),
  ).toEqual(
    expect.objectContaining({
      page: 1,
    }),
  );
  expect(paginatedResult.current.data).toEqual({
    ...fakeApiResult,
    result: [
      ...fakeApiResult.result.map(mapQueryResponse),
      ...fakeApiResult.result.map(mapQueryResponse),
    ],
  });
});
